### PR TITLE
Bulk "index" action and HEAD support

### PR DESCRIPTION
**Pull request by tcrossland**
This PR includes support for the bulk "index" action (create or replace), while leaving "create" as the default.

I've also included support for HTTP HEAD requests, which Elasticsearch uses in various places (index exists, type exists, etc.)

### CODE REVIEW

**Comment by tcrossland** [comment]
Another small improvement to allow index settings to be specified as a map instead of a file path (to work around issues like https://github.com/danielberkompas/elasticsearch-elixir/issues/51 and https://github.com/danielberkompas/elasticsearch-elixir/issues/76)

**Comment by tcrossland** [comment]
@danielberkompas thanks, I've changed the validation for `settings` to ensure that it's either a string (is_binary/1) or a map (is_map/1). I've removed the `presence` validation as it fails for empty maps (an empty map is now a valid value for `settings`)

**Comment by erikreedstrom** [comment]
It seems at least part of this functionality is covered under https://github.com/danielberkompas/elasticsearch-elixir/pull/68

We have a need to support delete actions as well in bulk, and would like to know which of the two PRs will be moving forward.

File: lib/elasticsearch/indexing/bulk.ex [hunk 1]
```diff
@@ -51,9 +51,9 @@ defmodule Elasticsearch.Index.Bulk do
       iex> Bulk.encode!(Cluster, 123, "my-index")
       ** (Protocol.UndefinedError) protocol Elasticsearch.Document not implemented for 123. This protocol is implemented for: Comment, Post
   """
-  def encode!(cluster, struct, index) do
+  def encode!(cluster, struct, index, action \\ "create") do
```
Comment (erikreedstrom) [inline lib/elasticsearch/indexing/bulk.ex]: It seems there should be an additional function head here for `"delete"`, which provides the header only.

> `delete` does not expect a source on the next line and has the same semantics as the standard delete API.
https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html

**Comment by erikreedstrom** [comment]
We've added some onto this in the expectation this is merged. :smile: We'll PR once this is in.

For reference, here's our changes compared to this branch: https://github.com/Bluetab/elasticsearch-elixir/compare/feature/bulk-index-action...carsdotcom:decouple_bulk_action

File: lib/elasticsearch/indexing/bulk.ex (hunk 1 above)
Comment (tcrossland) [reply lib/elasticsearch/indexing/bulk.ex]: @erikreedstrom seems reasonable to me. Our current use cases don't include bulk deletes so we're currently using the document delete API.

**Comment by tcrossland** [comment]
@danielberkompas is there anything preventing this being merged? thanks

**Comment by michelboaventura** [comment]
Hey @danielberkompas thank you for the great project, it's been really useful for us. Unfortunately I'm hitting the issue this changes seem to address. Does this PR need anything else to be merged?

Thanks!

**Comment by danielberkompas** [comment]
@michelboaventura nothing else was needed, I just lost track of it!

**Comment by michelboaventura** [comment]
Thank you @danielberkompas!

**Pull request merged**